android/src/main/java/org/mozilla/osmdroid/tileprovider/modules/TileDownloaderDelegate.java

```diff
@@ -52,80 +49,31 @@ public TileDownloaderDelegate(INetworkAvailablityCheck pNetworkAvailablityCheck,
     }
 
     /*
-     * Check if the tile is current by running a conditional get on
-     * the URL
+     * Write a tile from network to disk.
      */
-    public boolean isTileCurrent(SerializableTile serializableTile, ITileSource tileSource, MapTile tile) throws HttpHostConnectException {
+    public Drawable downloadTile(SerializableTile serializableTile, ITileSource tileSource, MapTile tile)
+            throws BitmapTileSourceBase.LowMemoryException {
         if (tileSource == null) {
-            return false;
+            Log.i(LOG_TAG, "tileSource is null");
+            return null;
         }
 
         if (networkIsUnavailable()) {
-            return false;
+            return null;
         }
 
         final String tileURLString = tileSource.getTileURLString(tile);
 
         if (tileURLString == null || tileURLString.length() == 0) {
-            return false;
+            return null;
         }
 
         if (urlIs404Cached(tileURLString)) {
-            // Note that this behaviour is different than
-            // actually downloading a tile.  If the URL is 404 cached,
-            // we just assume that the tile is 'current'
-            return true;
+            return null;
         }
 
         if (System.currentTimeMillis() < serializableTile.getCacheControl()) {
-            return true;
-        }
-
-        String etag = serializableTile.getEtag();
-        if (etag == null) {
-            // No etag means we want to download the file, no need
-            // to go checking the etag status over the network.
-            return false;
-        }
-
-        IHttpUtil httpUtil = new HttpUtil();
-        HashMap<String, String> headers = new HashMap<String, String>();
-        headers.put(ETAG_MATCH_HEADER, etag);
-        IResponse resp = httpUtil.get(tileURLString, headers);
-        if (resp == null) {
-            Log.w(LOG_TAG, "Error with network request.");
-            return false;
-        }
-
-        if (resp.httpResponse() == 304) {
-            // Resave the file - this will automatically update the cache-control value
-            serializableTile.saveFile();
-            return true;
-        }
-        return false;
-    }
-
-    /*
-     * Write a tile from network to disk.
-     */
-    public boolean downloadTile(ITileSource tileSource, MapTile tile) {
-        if (tileSource == null) {
-            Log.i(LOG_TAG, "tileSource is null");
-            return false;
-        }
-
-        if (networkIsUnavailable()) {
-            return false;
-        }
-
-        final String tileURLString = tileSource.getTileURLString(tile);
-
-        if (tileURLString == null || tileURLString.length() == 0) {
-            return false;
-        }
-
-        if (urlIs404Cached(tileURLString)) {
-            return false;
+            return tileSource.getDrawable(serializableTile.getTileData());
```

Comment:
I'd rather see that added in as a separate class and implemented as a proxy
to the delegate.

Separate the concerns and you can easily test that.  Load the image from
the tiledownloader,  if that fails,  try to load from disk.
On Jan 17, 2015 4:23 AM, "Christian Ascheberg" <notifications@github.com>
wrote:

> In
> android/src/main/java/org/mozilla/osmdroid/tileprovider/modules/TileDownloaderDelegate.java
> <https://github.com/mozilla/MozStumbler/pull/1400#discussion-diff-23124047>
> :
>
> > -            return false;
> > -        }
> > -
> > -        if (networkIsUnavailable()) {
> > -            return false;
> > -        }
> > -
> > -        final String tileURLString = tileSource.getTileURLString(tile);
> > -
> > -        if (tileURLString == null || tileURLString.length() == 0) {
> > -            return false;
> > -        }
> > -
> > -        if (urlIs404Cached(tileURLString)) {
> > -            return false;
> > +            return tileSource.getDrawable(serializableTile.getTileData());
>
> @crankycoder <https://github.com/crankycoder> Almost all return null in
> this method should be replaced by returning the cached tile, similar to
> this. Otherwise there is no map if offline etc.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/mozilla/MozStumbler/pull/1400/files#r23124047>.
>